File: src/Oisys.Web/wwwroot/js/models/order.js
(function (module) {
    module.factory("Order", [function () {
        function Order(id, code, customerId, date, dueDate, discountPercent, details) {
            this.id = id || 0;
            this.code = code || 0;
            this.customerId = customerId || 0;
            this.date = date || new Date();
            this.dueDate = dueDate || new Date();
            this.discountPercent = discountPercent || 0;
            this.details = details || [];

            this.grossAmount = 0;
            this.totalAmount = 0;
            this.discountAmount = 0;

            this.selectedCustomer = undefined;
        };

        Order.prototype = {
            update: function () {
                // Get customerId
                var priceListId = 1;

                if (this.selectedCustomer) {
                    this.customerId = this.selectedCustomer.id;
                    this.discountPercent = this.selectedCustomer.discount;
                    priceListId = this.selectedCustomer.priceListId;
                }
                
                var total = 0;
                if (this.details) {
                    this.details.forEach(function (elem) {
                        if (!elem.isDeleted) {
                            elem.setupDetail(priceListId);

                            if (elem.totalPrice) {
                                total += elem.totalPrice;
                            }
                        }
                    });
                }

                this.grossAmount = total;
                this.discountAmount = total * this.discountPercent / 100;
                this.totalAmount = total - this.discountAmount;
            }
        };

        return Order;
    }]);
})(angular.module("oisys-app"));

File: src/Oisys.Web/Models/Order.cs
﻿namespace Oisys.Web.Models
{
    using System;
    using System.Collections.Generic;
    using System.ComponentModel;
    using System.ComponentModel.DataAnnotations;
    using System.ComponentModel.DataAnnotations.Schema;
    using BlueNebula.Common.Helpers;

    /// <summary>
    /// <see cref="Order"/> class represents the Order object.
    /// </summary>
    public class Order : ModelBase, IObjectWithState
    {
        /// <summary>
        /// Gets or sets property Code.
        /// </summary>
        [Required]
        public int Code { get; set; }

        /// <summary>
        /// Gets or sets property CustomerId.
        /// </summary>
        [Required]
        public int CustomerId { get; set; }

        /// <summary>
        /// Gets or sets property Date.
        /// </summary>
        [Required]
        public DateTime Date { get; set; }

        /// <summary>
        /// Gets or sets property DueDate.
        /// </summary>
        public DateTime? DueDate { get; set; }

        /// <summary>
        /// Gets or sets property DiscountPercent.
        /// </summary>
        public decimal DiscountPercent { get; set; }

        /// <summary>
        /// Gets or sets property DiscountAmount.
        /// </summary>
        public decimal DiscountAmount { get; set; }

        /// <summary>
        /// Gets or sets property TotalAmount.
        /// </summary>
        public decimal TotalAmount { get; set; }

        /// <summary>
        /// Gets or sets a value indicating whether gets or sets property IsDeleted.
        /// </summary>
        [DefaultValue(false)]
        public bool IsDeleted { get; set; }

        /// <summary>
        /// Gets or sets property Customer navigation property.
        /// </summary>
        public Customer Customer { get; set; }

        /// <summary>
        /// Gets or sets property Details navigation property.
        /// </summary>
        public ICollection<OrderDetail> Details { get; set; }

        /// <summary>
        /// Gets or sets the objectstate property
        /// </summary>
        [NotMapped]
        public ObjectState State { get; set; }
    }
}